File: src/eqn/ast/EqnAstNodeSubtract.java
package eqn.ast;

import eqn.parser.exception.EqnException;

public class EqnAstNodeSubtract extends EqnAstNodeBinaryOperation {

    public EqnAstNodeSubtract(EqnAstNode leftOperand, EqnAstNode rightOperand) throws EqnException {
        super("-", leftOperand, rightOperand, PrecedenceType.Addition);
    }

    @Override
    public double evaluate() throws EqnException {
        return left().evaluate() - right().evaluate();
    }

    @Override
    public EqnAstNode simplify() throws EqnException {
        simplifyChildren();

        return new EqnAstNodeSetAdd(left(), new EqnAstNodeUnaryMinus(right())).simplify();
    }

}

File: src/eqn/ast/EqnAstNodeSubtract.kt
package eqn.ast

import eqn.ast.base.EqnAstNode
import eqn.ast.base.EqnAstNodeBinary
import eqn.parser.exception.EqnException

class EqnAstNodeSubtract(leftOperand: EqnAstNode, rightOperand: EqnAstNode) : EqnAstNodeBinary("-", Type.Subtraction, PrecedenceType.Addition, leftOperand, rightOperand) {
    @Throws(EqnException::class)
    override fun evaluate(arguments: Map<String, Double>?): Double {
        return left.evaluate(arguments) - right.evaluate(arguments)
    }

    @Throws(EqnException::class)
    override fun simplify(arguments: Map<String, Double>?): EqnAstNode {
        simplifyChildren(arguments)
        return EqnAstNodeAdd(left, EqnAstNodeUnaryMinus(right.simplify(arguments)))
    }
}